infrastructure-playbooks/shrink-rgw.yml

```diff
@@ -0,0 +1,138 @@
+
+    - import_role:
+        name: ceph-facts
+
+    - name: exit playbook, if no rgw was given
+      when: rgw_to_kill is not defined
+      fail:
+        msg: >
+          rgw_to_kill must be declared.
+          Exiting shrink-cluster playbook, no RGW was removed. On the command
+          line when invoking the playbook, you can use
+          "-e rgw_to_kill=ceph-rgw0 argument".  You can only remove a single
+          RGW each time the playbook runs.
+
+    - name: get rgw hostname
+      set_fact:
+        rgw_to_kill_hostname: "{{ rgw_to_kill.split('.')[0] }}"
+
+    # TODO: certainly, suffix is not the right name.
+    - name: get rgw suffix
+      set_fact:
+        rgw_to_kill_suffix: "{{ rgw_to_kill.split('.')[1] }}"
+
+    - name: exit if supplied hostname didnt match actual hostname
+      fail:
+        msg: supplied hostname didn't match with actual hostname
+      when: hostvars[rgw_to_kill_hostname]['ansible_hostname'] != rgw_to_kill_hostname
+
+    - name: exit playbook, if the rgw is not part of the inventory
+      when: rgw_to_kill_hostname not in groups[rgw_group_name]
+      fail:
+        msg: >
+          It seems that the host given is not part of your inventory, please
+          make sure it is.
+
+    - name: exit playbook, if user did not mean to shrink cluster
+      when: ireallymeanit != 'yes'
+      fail:
+        msg: >
+          Exiting shrink-mon playbook, no monitor was removed. To shrink the
+          cluster, either say 'yes' on the prompt or use
+          '-e ireallymeanit=yes' on the command line when  invoking the playbook
+
+    - name: set_fact container_exec_cmd for mon0
+      when: containerized_deployment | bool
+      set_fact:
+        container_exec_cmd: >
+          {{ container_binary }} exec ceph-mon-{{
+          hostvars[groups[mon_group_name][0]]['ansible_hostname'] }}
+
+    - name: exit playbook, if can not connect to the cluster
+      command: "{{ container_exec_cmd | default('') }} timeout 5 ceph --cluster {{ cluster }} health"
+      register: ceph_health
+      until: ceph_health.rc
+      failed_when: ceph_health.rc != 0
+      delegate_to: "{{ groups[mon_group_name][0] }}"
+      retries: 5
+      delay: 2
+
+  tasks:
+    - name: stop rgw service and verify it
+      block:
+        - name: stop rgw service
+          service:
+            name: ceph-radosgw@rgw.{{ rgw_to_kill }}
+            state: stopped
+            enabled: no
+          delegate_to: "{{ rgw_to_kill_hostname }}"
+          failed_when: false
+
+        - name: ensure that the rgw is stopped
+          command: "systemctl is-active ceph_rgw@{{ rgw_to_kill }}"
+          register: rgw_to_kill_status
+          failed_when: rgw_to_kill_status.rc == 0
+          delegate_to: "{{ rgw_to_kill_hostname }}"
+          retries: 5
+          delay: 2
+
+    # there were cases during writing this playbook that rgw_to_kill was spotted in json
+    # supplied by ceph -s
+    - name: sleep for 5 seconds
```

Comment:
I know this is kind of stuff pretty hard to handle, but don't we have something a bit more elegant?